--- a/src/bgpd/rtr_proto.c	2024-06-01 11:42:02.490253072 +0200
+++ b/src/bgpd/rtr_proto.c	2024-06-01 11:43:40.384138844 +0200
@@ -36,6 +36,7 @@
 } __packed;
 
 #define RTR_MAX_VERSION		2
+#define RTR_MIN_VERSION		1
 #define RTR_MAX_LEN		2048
 #define RTR_DEFAULT_REFRESH	3600
 #define RTR_DEFAULT_RETRY	600
@@ -1385,7 +1386,7 @@
 	rs->id = id;
 	rs->session_id = -1;
 	rs->version = RTR_MAX_VERSION;
-	rs->prev_version = RTR_MAX_VERSION;
+	rs->prev_version = RTR_MIN_VERSION;
 	rs->refresh = RTR_DEFAULT_REFRESH;
 	rs->retry = RTR_DEFAULT_RETRY;
 	rs->expire = RTR_DEFAULT_EXPIRE;
